File: frontend/types/conversation.ts
export type ConversationSummary = {
  id: number;
  title: string;
  created_at: string;
};

export type ConversationMessage = {
  id: number;
  role: "user" | "assistant";
  content: string;
  created_at: string;
};

export type ConversationDetail = ConversationSummary & {
  messages: ConversationMessage[];
};

export type ConversationTurn = {
  conversation_id: number;
  title: string;
  user_message: ConversationMessage;
  assistant_message: ConversationMessage;
};

export type ConversationCreateResponse = {
  conversation_id: number;
};

File: backend/models/conversation.py
"""Session 10: persisted conversations and ordered chat messages."""

from datetime import datetime, timezone

from sqlalchemy import CheckConstraint, Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from database import Base


def _utc_now() -> datetime:
    """Return an aware UTC timestamp for new conversation records."""

    return datetime.now(timezone.utc)


class Conversation(Base):
    """One private chat thread owned by one authenticated user."""

    __tablename__ = "conversations"

    id = Column(Integer, primary_key=True)
    user_id = Column(
        Integer,
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    title = Column(String(256), nullable=False)
    created_at = Column(DateTime(timezone=True), nullable=False, default=_utc_now)

    owner = relationship("User", back_populates="conversations")
    messages = relationship(
        "Message",
        back_populates="conversation",
        cascade="all, delete-orphan",
        passive_deletes=True,
        order_by="Message.id",
    )


class Message(Base):
    """One user or assistant turn in a conversation."""

    __tablename__ = "messages"
    __table_args__ = (
        CheckConstraint("role IN ('user', 'assistant')", name="ck_messages_role"),
    )

    id = Column(Integer, primary_key=True)
    conversation_id = Column(
        Integer,
        ForeignKey("conversations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    role = Column(String(16), nullable=False)
    content = Column(Text, nullable=False)
    created_at = Column(DateTime(timezone=True), nullable=False, default=_utc_now)

    conversation = relationship("Conversation", back_populates="messages")
